### PR TITLE
feat: full CTW3 support — PRD #25

### LINKED ISSUE #25: PRD: Full Eversweet Max 2 UVC (CTW3) Support with Scalable Device Architecture
## Problem Statement

As a Home Assistant user with a PetKit Eversweet Max 2 UVC (CTW3) water fountain, I want to fully control and monitor my device locally via BLE through Home Assistant, replacing the need for the PetKit iOS/Android app entirely. The current integration has partial CTW3 support but suffers from authentication conflicts with the official app, missing control entities (LED, DND schedules, child lock, settings), and lacks a scalable architecture for future PetKit BLE devices (feeders, litter boxes).

## Solution

Refactor the ha-petkit integration to:
1. **Fix authentication** by implementing device initialization (CMD 73) — HA claims the device with its own secret, stored persistently in the config entry. This replaces the fragile zero-secret approach and eliminates conflicts with the PetKit app.
2. **Add full device control** — LED brightness/schedule, DND schedule, smart mode timing, child lock, and UVC status via new HA entity platforms (switch, number, select, time).
3. **Add drink statistics tracking** — monitor pet drinking events based on pump activity and pet detection sensors.
4. **Refactor to a scalable device-class architecture** — abstract BLE protocol handling into a base class with device-specific subclasses, preparing for future support of feeders, litter boxes, and other PetKit BLE products.

## User Stories

1. As a HA user, I want the integration to initialize my CTW3 fountain with its own secret during setup, so that BLE authentication works reliably without conflicting with the PetKit app.
2. As a HA user, I want the device secret to be stored in the config entry, so that it survives HA restarts and I never lose access to my device.
3. As a HA user, I want to see a warning during setup that initializing the device will prevent the PetKit app from connecting, so that I can make an informed decision.
4. As a HA user, I want to turn my fountain on/off from HA, so that I can control it from automations and dashboards.
5. As a HA user, I want to switch between Normal, Smart, and Sensor modes, so that I can customize water flow behavior.
6. As a HA user, I want to adjust the smart mode work/sleep timing (minutes on, minutes off), so that I can fine-tune water usage.
7. As a HA user, I want to toggle the LED light on/off, so that I can control the indicator light.
8. As a HA user, I want to set the LED brightness level, so that I can dim or brighten the light.
9. As a HA user, I want to configure the LED on/off schedule (time-based), so that the light turns off at night automatically.
10. As a HA user, I want to toggle Do Not Disturb mode, so that the fountain runs silently during specified hours.
11. As a HA user, I want to set the DND start and end times, so that quiet hours match my schedule.
12. As a HA user, I want to toggle the child lock, so that physical button presses on the device are ignored.
13. As a HA user, I want to see the UVC disinfection status as a binary sensor, so that I know when UV sterilization is active.
14. As a HA user, I want to see the battery percentage and voltage, so that I know when to recharge.
15. As a HA user, I want to see whether the fountain is on AC power or battery, so that I know the power source.
16. As a HA user, I want to see the filter remaining percentage and estimated days, so that I know when to replace the filter.
17. As a HA user, I want to reset the filter counter via a button, so that the filter tracking restarts after replacement.
18. As a HA user, I want to see the pump runtime today and total, so that I can track device usage.
19. As a HA user, I want to see the estimated water purified today, so that I can monitor hydration.
20. As a HA user, I want to see the estimated energy usage today, so that I can track power consumption.
21. As a HA user, I want to see low water, filter, and breakdown warnings as binary sensors, so that I get alerted to problems.
22. As a HA user, I want to see the pet detection status, so that I know when my pet is near the fountain.
23. As a HA user, I want drink statistics (based on pump events and pet detection), so that I can track my pet's drinking habits.
24. As a HA user, I want the RSSI signal strength as a diagnostic sensor, so that I can troubleshoot BLE connectivity.
25. As a HA user, I want the firmware version as a diagnostic sensor, so that I can track device software.
26. As a HA user, I want the integration to support other PetKit water fountain models (W4X, W5, W5C, CTW2), so that one integration covers all fountain variants.
27. As a future HA user, I want the architecture to support adding PetKit feeders and litter boxes, so that the integration can grow to cover all PetKit BLE products.
28. As a developer, I want unit tests for protocol parsing (frame encode/decode, state parsing, auth flow), so that protocol changes don't silently break things.
29. As a HA user, I want the device to be auto-discovered via BLE advertisement matching, so that setup is seamless.
30. As a HA user, I want to write all settings to the device (CMD 221) in a single operation, so that changes are applied atomically.

### Authentication Overhaul
- During config flow (first setup), the integration will:
  1. Connect to the device and check if device_id == 0 (uninitialized)
  2. If uninitialized: generate a random 8-byte secret, send CMD 73 to initialize the device
  3. If already initialized: prompt the user to factory reset the device first
  4. Store the secret in `config_entry.data["secret"]` (hex-encoded)
- All subsequent connections use the stored secret for CMD 86 verification
- The config flow will show a warning that the PetKit app will no longer work after initialization

### Device-Class Architecture (Scalable for Future Devices)
- Create a base `PetkitBleDevice` abstract class with the common BLE protocol (frame encode/decode, connect/disconnect, auth)
- Create `PetkitFountainDevice` subclass for water fountains with fountain-specific commands and state parsing
- Future device types (feeders, litter boxes) will be separate subclasses
- Each device class defines its own `PetkitDeviceData` dataclass for parsed state
- Entity platforms use a registry pattern to map device class → entity descriptions

### New HA Entity Platforms
- **select.py**: Mode selection (Off / Normal / Smart / Sensor) via CMD 220
- **number.py**: Smart mode work minutes, smart mode sleep minutes, LED brightness via CMD 221
- **switch.py** (extended): Power switch (existing), LED toggle, DND toggle, child lock toggle via CMD 221
- **time.py** (new): LED schedule on/off times, DND schedule start/end times via CMD 225/226
- **binary_sensor.py** (extended): UVC status sensor (from module_status or dedicated field)
- **sensor.py** (extended): Drink event counter / frequency sensor

### Protocol Additions
- CMD 221 (Write Settings): smart timing, LED config, DND config, child lock — all in one payload
- CMD 225 (Write Light Schedule): LED on/off times
- CMD 226 (Write DND Schedule): DND start/end times
- CMD 73 with init payload: device initialization with secret
- Proper CMD response matching (verify response cmd matches request cmd)

### Secret Storage
- Secret stored as hex string in `config_entry.data`
- Migration path: existing config entries without a secret will trigger re-initialization flow

### What Makes a Good Test
- Tests should verify external behavior (input bytes → parsed data, command → expected frame bytes)
- Tests should NOT mock internal implementation details
- Tests should cover edge cases: short payloads, invalid frames, timeout scenarios

### Modules to Test
- **Protocol layer**: Frame encoding/decoding, all CMD parsers (210, 211, 66, 200, 213), settings payload building (221, 225, 226)
- **Authentication flow**: Zero-secret path, stored-secret path, initialization flow, error cases
- **State calculation**: Filter days remaining, water purified, energy consumption formulas
- **Device alias detection**: BLE name → alias mapping

### Test Infrastructure
- Use pytest with standard HA test patterns
- Mock BLE communication at the transport level
- No existing test suite — tests will be created from scratch in a `tests/` directory

## Out of Scope

- **PetKit Cloud/API integration** — this PRD covers BLE-only local control
- **Co-existence with PetKit iOS/Android app** — the integration claims exclusive device access
- **OTA firmware updates** (CMD 83) — too risky without official documentation
- **Specific feeder/litter box implementations** — only the architecture will be prepared; actual device support is a future PRD
- **Multi-device secret management UI** — secrets are stored automatically, no user-facing management needed

## Further Notes

- The CTW3 authentication quirk (always using `[0]*6` as device_id for secret computation) needs to be preserved for backward compatibility with devices that are already paired using this method
- Factory reset procedure for Eversweet Max 2 UVC: detach the control module from the device body, press and hold the power button for 3 seconds until all LEDs flash
- The reverse-engineering reference implementation at `github.com/mr-ransel/petkit-ble-reverse-engineering` is the primary protocol reference
- BLE name `Petkit_CTW3_100` corresponds to the Eversweet Max 2 UVC variant
- The integration should gracefully handle BLE disconnections mid-poll (common with Bluetooth proxies) and retry on the next polling interval

### PR DESCRIPTION
**Pull request by aavdberg**
## Summary

Implements PRD #25: full Home Assistant BLE integration for the PetKit Eversweet Max 2 UVC (CTW3).

### New features
- **Stored secret authentication** — device is initialized once during setup (CMD 73), secret stored persistently. Regular auth uses CMD 86 only.
- **Config flow initialization step** — detects uninitialized devices, generates secret, warns about iOS app exclusivity
- **LED switch & brightness** — toggle LED on/off, set brightness (1-10) via CMD 221
- **DND switch & schedule** — toggle Do Not Disturb, set start/end times
- **Child lock switch** — toggle child lock via CMD 221
- **Smart mode timing** — configure work/sleep minutes for Smart mode
- **LED schedule** — set LED on/off times
- **UVC binary sensor** — shows UVC sterilization status (CTW3)
- **Drink event counter** — tracks pet drinking events via detect_status edge detection

### New files
- \protocol.py\ — payload builders for CMD 73/84/220/221
- \
umber.py\ — smart work/sleep minutes, LED brightness entities
- \	ime.py\ — LED on/off and DND start/end schedule entities
- \	ests/\ — 39 unit tests (protocol builders, frame format, data model, parsers)

### Modified files
- \const.py\ — new CMDs (215/216/221/225/226), MODE constants
- \le_client.py\ — fixed auth flow, extended data model
- \config_flow.py\ — device init step with factory reset warning
- \coordinator.py\ — passes stored secret, drink event tracking
- \switch.py\ — LED, DND, child lock switches
- \inary_sensor.py\ — UVC active sensor
- \sensor.py\ — drink event count
- \__init__.py\ — NUMBER and TIME platforms
- All translation files (EN/NL/UK)

### Verification
- ✅ Ruff lint: all checks passed
- ✅ Tests: 39/39 passed

Closes #25

### CODE REVIEW

File: custom_components/petkit_ble/coordinator.py [hunk 1]
```diff
@@ -31,6 +31,15 @@ def __init__(self, hass: HomeAssistant, config_entry: ConfigEntry) -> None:
         self._config_entry = config_entry
         self._ble_lock = asyncio.Lock()
 
+        # Stored secret from device initialization (may be None for legacy entries)
+        secret_hex = config_entry.data.get(CONF_DEVICE_SECRET)
+        self._secret: bytes | None = bytes.fromhex(secret_hex) if secret_hex else None
```
[the remark this answers is not in the document]
Comment (aavdberg) [reply custom_components/petkit_ble/coordinator.py]: Fixed: wrapped `bytes.fromhex()` in try/except with warning log, treating corrupted secrets as None.

File: custom_components/petkit_ble/ble_client.py [hunk 2]
```diff
@@ -277,31 +290,24 @@ async def disconnect(self) -> None:
...
-        # CMD 73 — register the new random secret with the device
-        await self._send_and_wait(CMD_AUTH_INIT, FRAME_TYPE_SEND, list(device_id_be) + list(new_secret))
         await asyncio.sleep(AUTH_STEP_DELAY)
 
-        # CMD 86 — verify secret; response[0]==1 means success
-        payload_86 = await self._send_and_wait(CMD_AUTH_VERIFY, FRAME_TYPE_SEND, list(new_secret))
+        # CMD 86 — verify with stored secret
+        if secret is None:
+            secret = bytes(8)  # zero secret for uninitialized devices
+        payload_86 = await self._send_and_wait(CMD_AUTH_VERIFY, FRAME_TYPE_SEND, list(secret[:8]))
```
[the remark this answers is not in the document]
Comment (aavdberg) [reply custom_components/petkit_ble/ble_client.py:311]: Addressed: The stored-secret approach is intentional. With device initialization (CMD 73), we write a known secret to the device during setup. CMD 86 with that stored secret is a genuine auth — not a false positive. The false-positive issue only occurred with random/zero secrets where CMD 86 appeared to succeed but the device ignored subsequent commands. Since we now persist the real secret from CMD 73, CMD 86 success is reliable. For legacy entries without a secret, the zero-secret fallback is maintained (backward compatible). If a device is factory reset, auth will fail and the user needs to re-add the device.

File: custom_components/petkit_ble/config_flow.py [hunk 3]
```diff
@@ -161,6 +162,82 @@ async def async_step_user(self, user_input: dict[str, Any] | None = None) -> Con
...
+            # a previous installation, or the device accepts zero-secret).
+            errors["base"] = "device_already_initialized"
+            if user_input is not None:
+                # User acknowledged — create entry without secret
+                return self.async_create_entry(title=name, data=self._pending_data)
+            return self.async_show_form(
+                step_id="init_device",
+                description_placeholders={"name": name},
+                errors=errors,
+            )
```
[the remark this answers is not in the document]
Comment (aavdberg) [reply custom_components/petkit_ble/config_flow.py]: Fixed: Now aborts with `device_already_initialized` reason, forcing factory reset. Added abort strings to all translation files.

File: custom_components/petkit_ble/config_flow.py [hunk 4]
```diff
@@ -161,6 +162,82 @@ async def async_step_user(self, user_input: dict[str, Any] | None = None) -> Con
...
+                    step_id="init_device",
+                    description_placeholders={"name": name},
+                    errors=errors,
+                )
+            client2 = PetkitBleClient(ble_device2)
+            await client2._connect()
+            try:
+                success = await client2.async_init_device(device_id, secret)
+            finally:
+                await client2.disconnect()
```
[the remark this answers is not in the document]
Comment (aavdberg) [reply custom_components/petkit_ble/config_flow.py]: Fixed: `async_init_device()` now handles its own connect/disconnect (mirroring `async_check_initialized()`). No more private `_connect()` usage from config flow.

File: custom_components/petkit_ble/time.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        field_name="dnd_start_minutes",
+        available_fn=lambda d: d.alias not in CTW3_ALIASES,
+    ),
+    PetkitTimeDescription(
+        key="dnd_end_time",
+        translation_key="dnd_end_time",
+        value_fn=lambda d: d.dnd_end_minutes,
+        field_name="dnd_end_minutes",
+        available_fn=lambda d: d.alias not in CTW3_ALIASES,
+    ),
```
[the remark this answers is not in the document]
Comment (aavdberg) [reply custom_components/petkit_ble/time.py:71]: This is intentional. CTW3 CMD 211 response does not include LED/DND schedule times (only on/off toggle flags). The CTW3 controls LED and DND via simple switches (CMD 221), not time-based schedules. CMD 225/226 constants are defined for future use when the protocol is fully reverse-engineered. Time entities are correctly disabled for CTW3.

File: custom_components/petkit_ble/number.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        smart_work=overrides.get("smart_time_on", data.smart_time_on),
+        smart_sleep=overrides.get("smart_time_off", data.smart_time_off),
+        led_switch=overrides.get("led_switch", data.led_switch),
+        led_brightness=overrides.get("led_brightness", data.led_brightness),
+        led_on_minutes=overrides.get("led_on_minutes", data.led_on_minutes),
+        led_off_minutes=overrides.get("led_off_minutes", data.led_off_minutes),
+        dnd_enabled=overrides.get("do_not_disturb_switch", data.do_not_disturb_switch),
+        dnd_start_minutes=overrides.get("dnd_start_minutes", data.dnd_start_minutes),
+        dnd_end_minutes=overrides.get("dnd_end_minutes", data.dnd_end_minutes),
+        child_lock=overrides.get("is_locked", data.is_locked),
```
[the remark this answers is not in the document]
Comment (aavdberg) [reply custom_components/petkit_ble/number.py]: Fixed: Centralized `build_full_settings_payload()` in `protocol.py` as a single shared helper. All three platforms (switch, number, time) now import from there. Private `_battery_work_time`/`_battery_sleep_time` fields made public.

File: tests/test_data_model.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+
+    def test_normal_mode_full(self) -> None:
+        """100% filter in normal mode → ceil((1.0 * 30) * 1/1) = 30 days."""
+        data = PetkitFountainData(mode=1, filter_percent=100)
+        assert data.filter_days_remaining == 30
+
+    def test_normal_mode_half(self) -> None:
+        """50% filter in normal mode → ceil((0.5 * 30) * 1/1) = 15 days."""
+        data = PetkitFountainData(mode=1, filter_percent=50)
+        assert data.filter_days_remaining == 15
```
[the remark this answers is not in the document]
Comment (aavdberg) [reply tests/test_data_model.py]: Fixed: Normal mode now uses `FILTER_LIFE_NORMAL_DAYS` (60) directly. Tests updated: 100% = 60 days, 50% = 30 days.

File: custom_components/petkit_ble/protocol.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        13,
+    ]
+
+
+def build_init_payload(device_id: int, secret: bytes) -> list[int]:
+    """Build the 16-byte payload for CMD 73 (device init).
+
+    Layout: 8-byte device_id (big-endian) + 8-byte secret.
+    """
+    id_bytes = list(struct.pack(">q", device_id))
```
[the remark this answers is not in the document]
Comment (aavdberg) [reply custom_components/petkit_ble/protocol.py]: Fixed: Changed to `>Q` (unsigned) for full 64-bit device ID support.

File: custom_components/petkit_ble/sensor.py [hunk 9]
```diff
@@ -129,6 +129,13 @@ class PetkitSensorEntityDescription(SensorEntityDescription):
         state_class=SensorStateClass.MEASUREMENT,
         value_fn=lambda d: d.rssi,
     ),
+    PetkitSensorEntityDescription(
+        key="drink_count",
+        translation_key="drink_count",
+        state_class=SensorStateClass.TOTAL_INCREASING,
```
[the remark this answers is not in the document]
Comment (aavdberg) [reply custom_components/petkit_ble/sensor.py]: Fixed: Changed to `SensorStateClass.MEASUREMENT` since the counter is in-memory only and resets on restart.

File: custom_components/petkit_ble/coordinator.py [hunk 10]
```diff
@@ -52,14 +61,28 @@ async def _async_update_data(self) -> PetkitFountainData:
...
+        if (
+            self._prev_detect_status is not None
+            and self._prev_detect_status == 0
+            and cur_detect == 1
+            and cur_pump
+        ):
+            self._drink_event_count += 1
+        self._prev_detect_status = cur_detect
+        self._prev_pump_running = cur_pump
+        data.drink_event_count = self._drink_event_count
```
[the remark this answers is not in the document]
Comment (aavdberg) [reply custom_components/petkit_ble/coordinator.py:81]: Fixed: Removed `_prev_pump_running` — it was unused leftover state.

File: custom_components/petkit_ble/time.py [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+from .entity import PetkitBleEntity
+from .protocol import build_settings_payload_ctw3, build_settings_payload_generic
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _minutes_to_time(minutes: int) -> datetime.time:
+    """Convert minutes-from-midnight to a time object."""
+    h = min(minutes // 60, 23)
+    m = minutes % 60
```
[the remark this answers is not in the document]
Comment (aavdberg) [reply custom_components/petkit_ble/time.py]: Fixed: Now uses `minutes % 1440` for safe wrapping of out-of-range values.

File: custom_components/petkit_ble/switch.py [hunk 12]
```diff
@@ -20,14 +22,46 @@
...
+        smart_sleep=overrides.get("smart_time_off", data.smart_time_off),
+        led_switch=overrides.get("led_switch", data.led_switch),
+        led_brightness=overrides.get("led_brightness", data.led_brightness),
+        led_on_minutes=overrides.get("led_on_minutes", data.led_on_minutes),
+        led_off_minutes=overrides.get("led_off_minutes", data.led_off_minutes),
+        dnd_enabled=overrides.get("do_not_disturb_switch", data.do_not_disturb_switch),
+        dnd_start_minutes=overrides.get("dnd_start_minutes", data.dnd_start_minutes),
+        dnd_end_minutes=overrides.get("dnd_end_minutes", data.dnd_end_minutes),
+        child_lock=overrides.get("is_locked", data.is_locked),
+    )
```
[the remark this answers is not in the document]
Comment (aavdberg) [reply custom_components/petkit_ble/switch.py]: Fixed: Removed duplicate, now imports shared `build_full_settings_payload` from `protocol.py`. Battery timing fields are public.

File: custom_components/petkit_ble/time.py [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+        smart_work=overrides.get("smart_time_on", data.smart_time_on),
+        smart_sleep=overrides.get("smart_time_off", data.smart_time_off),
+        led_switch=overrides.get("led_switch", data.led_switch),
+        led_brightness=overrides.get("led_brightness", data.led_brightness),
+        led_on_minutes=overrides.get("led_on_minutes", data.led_on_minutes),
+        led_off_minutes=overrides.get("led_off_minutes", data.led_off_minutes),
+        dnd_enabled=overrides.get("do_not_disturb_switch", data.do_not_disturb_switch),
+        dnd_start_minutes=overrides.get("dnd_start_minutes", data.dnd_start_minutes),
+        dnd_end_minutes=overrides.get("dnd_end_minutes", data.dnd_end_minutes),
+        child_lock=overrides.get("is_locked", data.is_locked),
```
[the remark this answers is not in the document]
Comment (aavdberg) [reply custom_components/petkit_ble/time.py]: Fixed: Removed duplicate, imports shared helper from `protocol.py`.

File: tests/test_protocol.py [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+        """Payload must be 16 bytes (8 device_id + 8 secret)."""
+        result = build_init_payload(0, bytes(8))
+        assert len(result) == 16
+
+    def test_device_id_encoding(self) -> None:
+        """Device ID is packed big-endian as a signed 64-bit int."""
+        result = build_init_payload(12345, b"\x01\x02\x03\x04\x05\x06\x07\x08")
+        id_bytes = bytes(result[:8])
+        device_id = struct.unpack(">q", id_bytes)[0]
+        assert device_id == 12345
```
[the remark this answers is not in the document]
Comment (aavdberg) [reply tests/test_protocol.py]: Fixed: Test updated to use `>Q` (unsigned) matching the production code fix.
